go/obscuronode/enclave/sql/batch.go

```diff
@@ -0,0 +1,103 @@
+package sql
+
+import (
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethdb"
+)
+
+// ---- Implement the geth Batch interface, re-using ideas and types from geth's memorydb.go ----
+
+// keyvalue is a key-value tuple that can be flagged with a deletion field to allow creating database write batches.
+type keyvalue struct {
+	key    []byte
+	value  []byte
+	delete bool
+}
+
+type sqlBatch struct {
+	db     *sqlEthDatabase
+	writes []keyvalue
+	size   int
+}
+
+// Put inserts the given value into the batch for later committing.
+func (b *sqlBatch) Put(key, value []byte) error {
+	b.writes = append(b.writes, keyvalue{common.CopyBytes(key), common.CopyBytes(value), false})
+	b.size += len(key) + len(value)
+	return nil
+}
+
+// Delete inserts the a key removal into the batch for later committing.
+func (b *sqlBatch) Delete(key []byte) error {
+	b.writes = append(b.writes, keyvalue{common.CopyBytes(key), nil, true})
+	b.size += len(key)
+	return nil
+}
+
+// ValueSize retrieves the amount of data queued up for writing.
+func (b *sqlBatch) ValueSize() int {
+	return b.size
+}
+
+// Write executes a batch statement with all the updates
+func (b *sqlBatch) Write() error {
+	tx, err := b.db.db.Begin()
+	if err != nil {
+		return fmt.Errorf("failed to prepare batch transaction - %w", err)
+	}
+	pIns, err := tx.Prepare(`insert or replace into kv values(?, ?);`)
```

Comment:
aren't these the same prepared statements used by the non-batch stuff?

Reply:
Ah that's a good point. I thought it had to be duplicated because this is a prepared statement for use in the context of batch transaction.

But the `Tx` type has a method `tx.Stmt(stmt)` that you can give an existing prepared statement and it will give you a contextual version.